src/tree/updater_quantile_hist.cc

```diff
@@ -549,290 +1055,113 @@ void QuantileHistMaker::Builder::InitData(const GHistIndexMatrix& gmat,
+    const int32_t  nid         = nodes[node_idx].nid;
+    const int32_t  sibling_nid = nodes[node_idx].sibling_nid;
+    const int32_t  parent_nid  = nodes[node_idx].parent_nid;
+
+    common::GradStatHist::GradType* hist_data =
+        reinterpret_cast<common::GradStatHist::GradType*>(hist_[nid].data());
+    common::GradStatHist::GradType* sibling_hist_data = sibling_nid > -1 ?
+        reinterpret_cast<common::GradStatHist::GradType*>(
+          hist_[sibling_nid].data()) : nullptr;
+    common::GradStatHist::GradType* parent_hist_data  = sibling_nid > -1 ?
+        reinterpret_cast<common::GradStatHist::GradType*>(hist_[parent_nid].data()) : nullptr;
+
+    // reduce needed part of a hist here to have it in cache before enumeration
+    if (!rabit::IsDistributed()) {
+      const std::vector<uint32_t>& cut_ptr = gmat.cut.row_ptr;
+      const size_t ibegin = 2 * cut_ptr[fid];
+      const size_t iend = 2 * cut_ptr[fid + 1];
+      ReduceHistograms(hist_data, sibling_hist_data, parent_hist_data, ibegin, iend, node_idx,
+          hist_is_init, hist_buffers);
+    }
 
-  row_set_collection_.AddSplit(
-      nid, row_split_tloc_, (*p_tree)[nid].LeftChild(), (*p_tree)[nid].RightChild());
-  builder_monitor_.Stop("ApplySplit");
-}
+    if (spliteval_->CheckFeatureConstraint(nid, fid)) {
+      auto& snode = snode_[nid];
+      const bool compute_backward = this->EnumerateSplit(+1, gmat, hist_[nid], snode,
+          info, &splits[i].first, fid, nid);
 
-void QuantileHistMaker::Builder::ApplySplitDenseData(
-    const RowSetCollection::Elem rowset,
-    const GHistIndexMatrix& gmat,
-    std::vector<RowSetCollection::Split>* p_row_split_tloc,
-    const Column& column,
-    bst_int split_cond,
-    bool default_left) {
-  std::vector<RowSetCollection::Split>& row_split_tloc = *p_row_split_tloc;
-  constexpr int kUnroll = 8;  // loop unrolling factor
-  const size_t nrows = rowset.end - rowset.begin;
-  const size_t rest = nrows % kUnroll;
-
-#pragma omp parallel for num_threads(nthread_) schedule(static)
-  for (bst_omp_uint i = 0; i < nrows - rest; i += kUnroll) {
-    const bst_uint tid = omp_get_thread_num();
-    auto& left = row_split_tloc[tid].left;
-    auto& right = row_split_tloc[tid].right;
-    size_t rid[kUnroll];
-    uint32_t rbin[kUnroll];
-    for (int k = 0; k < kUnroll; ++k) {
-      rid[k] = rowset.begin[i + k];
-    }
-    for (int k = 0; k < kUnroll; ++k) {
-      rbin[k] = column.GetFeatureBinIdx(rid[k]);
-    }
-    for (int k = 0; k < kUnroll; ++k) {                      // NOLINT
-      if (rbin[k] == std::numeric_limits<uint32_t>::max()) {  // missing value
-        if (default_left) {
-          left.push_back(rid[k]);
-        } else {
-          right.push_back(rid[k]);
-        }
-      } else {
-        if (static_cast<int32_t>(rbin[k] + column.GetBaseIdx()) <= split_cond) {
-          left.push_back(rid[k]);
-        } else {
-          right.push_back(rid[k]);
-        }
+      if (compute_backward) {
```

Comment:
```suggestion
      // Sometimes, we don't need to enumerate backward because forward and backward
      // enumeration will give same loss values. This is the case if the particular feature 
      // column contains no missing values. So enumerate backward only if it's necessary.
      if (compute_backward) {
```